test_util/cluster.py

```diff
@@ -0,0 +1,260 @@
+            # key must be chmod 600 for test_runner to use
+            os.chmod(self.ssh_key_path, stat.S_IREAD | stat.S_IWRITE)
+        except FileNotFoundError as exc:
+            raise Exception('SSH key not found at path {}'.format(self.ssh_key_path)) from exc
+        except OSError as exc:
+            raise Exception('Unable to set permissions on SSH key at {}: {}'.format(
+                self.ssh_key_path,
+                exc,
+            )) from exc
+
+        self.bootstrap_host, self.masters, self.agents, self.public_agents = (
+            self.partition_cluster(
+                self.get_cluster_hosts(public_ips, setup),
+                num_masters,
+                num_agents,
+                num_public_agents,
+            )
+        )
+
+        self.test_host = self.masters[0]
+        logging.info('Test host public/private IP: ' + self.test_host.public_ip + '/' + self.test_host.private_ip)
+
+    @classmethod
+    def from_ccm_vpc(
+        cls,
+        cluster_name_prefix,
+        ssh_key_path,
+        num_masters,
+        num_agents,
+        num_public_agents,
+        region='us-west-2',
+        prereqs_installed=True,
+        setup=True
+    ):
+        cluster_name = cluster_name_prefix + '-' + ''.join(
+            random.choice(string.ascii_uppercase + string.digits) for _ in range(10)
+        )
+        if prereqs_installed:
+            os_name = 'cent-os-7-dcos-prereqs'
+        else:
+            os_name = 'cent-os-7'
+
+        logging.info('Creating AWS VPC cluster {}'.format(cluster_name))
+        vpc = test_util.ccm.Ccm().create_vpc(
+            name=cluster_name,
+            time=60,
+            instance_count=(1 + num_masters + num_agents + num_public_agents),
+            instance_type='m4.xlarge',
+            instance_os=os_name,
+            region=region,
+            key_pair_name=cluster_name,
+        )
+        logging.info('SSH key for cluster {}: {}'.format(cluster_name, vpc.get_ssh_key_url()))
+        # Write out the ssh key to the local filesystem for the ssh lib to pick up.
+        with open(ssh_key_path, 'w') as ssh_key_fh:
+            ssh_key_fh.write(vpc.get_ssh_key())
+
+        cluster = cls(
+            vpc.hosts(),
+            ssh_user='centos',
+            ssh_key_path=ssh_key_path,
+            remote_dir='/home/centos',
+            num_masters=num_masters,
+            num_agents=num_agents,
+            num_public_agents=num_public_agents,
+            setup=setup,
+        )
+        cluster.vpc = vpc
+        return cluster
+
+    def cleanup(self):
+        if self.vpc is not None:
+            self.vpc.delete()
+
+    @staticmethod
+    def partition_cluster(hosts, num_masters, num_agents, num_public_agents):
+        """Return (bootstrap, masters, agents, public_agents) from hosts."""
+        assert num_masters in (1, 3, 5)
+        assert all(x >= 0 for x in [num_agents, num_public_agents])
+        assert num_agents + num_public_agents > 0
```

Comment:
the requirements hard coded into the tests are currently
1. num_agents >= 2
2. num_public_agents >= 1